### ISSUE TITLE
Coverity report: Memory corruption in deflate_rle

### ISSUE THREAD

**Issue by Dead2** [zlib-ng/zlib-ng#1073]
This potential bug was found in develop but not in stable:
```
*** CID 366706:  Memory - corruptions  (OVERRUN)
/deflate_rle.c: 58 in deflate_rle()
52             }
53     
54             /* Emit match if have run of STD_MIN_MATCH or longer, else emit literal */
55             if (match_len >= STD_MIN_MATCH) {
56                 check_match(s, s->strstart, s->strstart - 1, match_len);
57     
>>>     CID 366706:  Memory - corruptions  (OVERRUN)
>>>     Overrunning callee's array of size 256 by passing argument "match_len - 3U" (which evaluates to 256) in call to "zng_tr_tally_dist".
58                 bflush = zng_tr_tally_dist(s, 1, match_len - STD_MIN_MATCH);
59     
60                 s->lookahead -= match_len;
61                 s->strstart += match_len;
62                 match_len = 0;
63             } else {
```

**Comment by nmoinvaz** [comment]
I'm not so sure this is a real issue. `match_len` can't be larger than `STD_MAX_MATCH`. Coverity may be making some wrong assumptions:

```
    match_len = STD_MAX_MATCH - (unsigned int)(strend - scan);
cond_at_least: Checking match_len > s->lookahead implies that match_len is at least 260 on the true branch.
assignment: Assigning: match_len = (match_len > s->lookahead) ? s->lookahead : match_len. The value of match_len is now at least 259.
    match_len = MIN(match_len, s->lookahead);
```

**Comment by Dead2** [comment]
I am also thinking this is not a real issue, but is there a way we can make that clear to Coverity (and other static checkers) in a good way?

I think most/all of them accept assert checks as ground truth, so that might be a possibility (and could also provide sometimes useful checkpoints against things becoming completely out of whack during code changes).

Such asserts might also improve static checking as they can define narrower limits than INT_MAX for example, possibly something to consider using more of in general, especially for functions where we trust the input to be within limits and don't actually check them like we would with exported functions.

**Comment by Dead2** [comment]
Just ignoring these kinds of warnings will only end up with us answering questions about them time and again, just like zlib does.

**Comment by mtl1979** [comment]
It clearly assumes `strend - scan` results in negative number, which is pointless as it's cast to `unsigned int` which would result in either positive number or overflow.

**Comment by nmoinvaz** [comment]
> I think most/all of them accept assert checks as ground truth, so that might be a possibility (and could also provide sometimes useful checkpoints against things becoming completely out of whack during code changes).

@Dead2 we check the length of the match in `check_match`. Do you think we should also add assert statements in `check_match`?

Maybe it is not getting checked because of `ZLIB_DEBUG` is not defined. We could do `#if defined(ZLIB_DEBUG || defined(__COVERITY__)`..

**Comment by Dead2** [comment]
Hmm, seems my reply was lost.
Adding coverity to the ifdef check did not improve anything unfortunately.

Not sure how to make this more clear to static checkers

**Comment by mtl1979** [comment]
I'm pretty sure static checkers can't handle sign cast using unnamed temporary variable... Even normal compilers consider them dead code and free to optimize away...

**Comment by nmoinvaz** [comment]
@Dead2 I'm surprised that `|| defined(__COVERITY__)` did not work for defining `check_match`. This might do better, to define `check_match` as assert statements when coverity is detected.

```c
#ifdef ZLIB_DEBUG
...
#elif defined(__COVERITY__)
#define check_match(s, start, match, length) \
    Assert(length >= STD_MIN_MATCH && length <= STD_MAX_MATCH, "invalid match length"); \
    Assert(match != start, "invalid match position"); \
#else
#define check_match(s, start, match, length)
#endif
```
If that doesn't fix it, we should just mark it as false positive in coverity and move on..

**Comment by Dead2** [comment]
I even tried changing it to
```C
#ifdef ZLIB_DEBUG
...
#else
#define check_match(s, start, match, length) \
    assert(length >= STD_MIN_MATCH && length <= STD_MAX_MATCH); \
    assert(match != start);
#endif
```

That still did not help on the warnings in Coverity, and I think it really should have. Are we overlooking anything?

**Comment by Dead2** [comment]
Fixed by #1476
